simx86: use lazy flags exclusively between instructions in codegen-sim.c

bartoldeman opened the pull request:
This is all borrowed from Bochs and then QEMU (who borrowed from Bochs making a few modifications).

Internally `codegen-sim` now only uses lazy flags: they get set entering Exec_sim and re-translated to EFLAGS on leaving Exec_sim, so there is no more need for RFL.valid, as they are always valid.

There are obviously some instructions need the non-lazy `EFLAGS`: `pushf`, `lahf`, and `sahf`; they now always translate to and from the lazy flags. Also `bsf` and `bsr` are a little unusual in only affecting ZF (though other flags are undocumented), so I implemented that the expensive way for now translating to EFLAGS, setting ZF and back from EFLAGS to lazy.

There is still some room for various optimizations here.

bartoldeman committed:
simx86, Gen_sim: make CF lazy where other flags are

bartoldeman committed:
simx86, Gen_sim: use lazy flags in O_CLEAR, O_SBSELF, O_SETCC

bartoldeman committed:
simx86, Gen_sim: always use lazy flags for CF,OF,ZF

Makes for quicker checks though needs setup in Exec_sim.

bartoldeman committed:
simx86, Gen_sim: also always use lazy flags for AF

Since RFL.valid=V_GEN no longer signals the use EFLAGS.AF directly,
make sure to preserve the lazy AF in RFL.cout for instructions that
set V_GEN.

bartoldeman committed:
simx86, Gen_sim: lazy flag symbolic constants, simplify FlagHandle* bitshifts

Same notation as in Bochs and QEMU; we don't care about cout bits 8-29 so
they don't need to be shifted away.

bartoldeman committed:
simx86, Gen_sim: use sign delta flag in RFL.cout

This is used instead of the sign bit of RFL.res only if the flags come
directly from EFLAGS; in that case the RFL.res sign bit is always 0
(so we can use either "or" or "xor").

bartoldeman committed:
simx86, Gen_sim: use parity delta flag in RFL.cout

This is used exactly like the sign delta bit, except the RFL.res parity when
set directly from EFLAGS is 1, so PD is flipped when setting it from EFLAGS.PF.

bartoldeman committed:
simx86, Gen_sim: remove now unused RFL.valid

stsp commented on src/base/emu-i386/simx86/codegen-sim.c:
```diff
@@ -2567,7 +2577,7 @@ unsigned int Gen_sim(const IGen *IG)
 		case CMC:
 			GTRACE0("O_CMC");
-			CPUBYTE(Ofs_FLAGS) ^= 1;
+			SET_CF(is_cf_set()^1);
```
Is this an optimization?
Maybe you need xor_cf(),
which will definitely be faster?

bartoldeman replied:
this was part of my "there is still room for optimizations". But this one is so simple to do with lazy flags that I could just inline it, so I did it just now.

bartoldeman committed:
simx86, Gen_sim: optimize CMC

This can be easily done with lazy flags by flipping both CF and PO in RFL.cout.

stsp commented:
Ok, thanks. :)

bartoldeman commented:
Apart from some small optimizations here and there I expect most of the potential speed-ups to come from two strategies:
1. Merge Exec_sim and Gen_sim into a single function with local variables for DR1, AR1, etc so the compiler can optimize them; their state needs to be preserved between intermediate ops in a single instruction but not between instructions. That means rewriting some external callers of Gen_sim to use Exec_sim instead.
2. Compile Exec_sim twice, one with debugging, one without (which can then also use -O2, and disable ubsan). This way not every single instruction needs to check `debug_level('e')`, we'll just invoke `Exec_sim_debug` if runtime debugging is enabled.

stsp commented:
> Merge Exec_sim and Gen_sim into a single function

Doesn't this bring us to an
old design, where sim could
only immediately exec (w/o
actual gen)?

Anyway, another source of
improvement may be to split
the interp loop to PM and RM
parts, as this will avoid quite
a lot of checks (caller always
knows the mode). Although
that's probably not a small
task.

bartoldeman commented:
> Doesn't this bring us to an
> old design, where sim could
> only immediately exec (w/o
> actual gen)?

No, it'd just move the loop and the big switch statement in the same function so local variables are preserved between loop iterations.

In my own testing with perf, Interp isn't a bottleneck at all any more (like 50x less than Gen_sim), though perhaps it's a bottleneck for you somewhere (with JIT?)

stsp commented:
Well, I guess its not a
bottleneck now with your
changes, but it definitely
was before. :)
Note that you already over-speed
dosbox on zx-emul by a
few orders of magnitude,
so that's a great work
(and that's when I compare
our sim to dosbox's jit!).
But for example in GTA
in hi-color mode we are
still too slow. To test GTA
you need to first revert
1ad9d735b as otherwise
GTA will keep polling the
retrace register.

stsp commented:
Please fix the SINGLESTEP build,
as its currently broken.
